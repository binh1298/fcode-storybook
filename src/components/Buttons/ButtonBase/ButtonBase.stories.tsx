import { Add as AddIcon, Edit as EditIcon } from "@material-ui/icons";

import ButtonBase, { ButtonBaseProps } from ".";

import { Story } from "@storybook/react";

//üëá This default export determines where your story goes in the story list
export default {
    title: "Components/ButtonBase",
    component: ButtonBase,
};

//üëá We create a ‚Äútemplate‚Äù of how args map to rendering
const Template: Story<ButtonBaseProps> = (args) => <ButtonBase {...args} />;

export const Primary = Template.bind({});
Primary.args = {
    children: "Compose",
    color: "primary",
    startIcon: <AddIcon />,
    style: { borderRadius: "50em" },
    variant: "contained",
};

export const Secondary = Template.bind({});
Secondary.args = {
    children: "Compose",
    color: "secondary",
    startIcon: <EditIcon />,
    style: { borderRadius: "50em" },
    variant: "contained",
};

export const FabPrimary = Template.bind({});
FabPrimary.args = {
    children: <AddIcon />,
    color: "primary",
    style: {
        height: "60px",
        width: "60px",
        borderRadius: "50%",
    },
    variant: "contained",
};

export const FabSecondary = Template.bind({});
FabSecondary.args = {
    children: <EditIcon />,
    color: "secondary",
    style: {
        height: "60px",
        width: "60px",
        borderRadius: "50%",
    },
    variant: "contained",
};
