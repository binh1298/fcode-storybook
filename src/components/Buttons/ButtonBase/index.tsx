import { Button as MaterialButton, ButtonProps as MaterialButtonProps } from "@material-ui/core";

import BoxBase from "../../Boxs/BoxBase";

export interface ButtonBaseProps extends MaterialButtonProps {
    pl?: number | undefined;
    pr?: number | undefined;
    pt?: number | undefined;
    pb?: number | undefined;
}

const buttonStyle = {
    borderRadius: "calc(1rem + 6px)",
};

const ButtonBase = (props: ButtonBaseProps) => {
    const { pl, pr, pt, pb, ...rest } = props;

    return (
        <BoxBase pl={pl} pr={pr} pt={pt} pb={pb}>
            <MaterialButton {...rest} style={buttonStyle} />
        </BoxBase>
    );
};

export default ButtonBase;
