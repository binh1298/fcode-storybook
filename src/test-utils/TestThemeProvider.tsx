import { PropsWithChildren } from "react";

import { ThemeProvider, StyledEngineProvider } from "@mui/material/styles";

import replyTheme from "src/theme/replyTheme";

const TestThemeProvider = (props: { children: PropsWithChildren<{}> }) => {
    return (
        <StyledEngineProvider injectFirst>
            <ThemeProvider theme={replyTheme}>{props.children}</ThemeProvider>
        </StyledEngineProvider>
    );
};

export default TestThemeProvider;
