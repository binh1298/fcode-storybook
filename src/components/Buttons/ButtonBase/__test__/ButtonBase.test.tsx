import ButtonBase, { ButtonBaseProps } from "..";

import { render, screen } from "@testing-library/react";
import TestThemeProvider from "src/test-utils/TestThemeProvider";

describe("<ButtonBase />", () => {
    let props: ButtonBaseProps;
    const primaryColor = "#344955";
    const secondaryColor = "#F9AA33";

    it("should exist children with primary color", () => {
        props = {
            children: "Test",
        };

        render(
            <TestThemeProvider>
                <div>
                    <ButtonBase {...props} />
                </div>
            </TestThemeProvider>
        );

        const muiButton = screen.getByTestId("ButtonBase__root");
        expect(muiButton).not.toBeDisabled();
        expect(screen.getByText("Test")).toBeInTheDocument();
    });

    it("should exist children with isLoading true", () => {
        props = {
            children: "Test",
            isLoading: true,
        };

        render(
            <TestThemeProvider>
                <ButtonBase {...props} />
            </TestThemeProvider>
        );

        const muiButton = screen.getByTestId("ButtonBase__root");
        const muiCircularProgress = screen.getByTestId("CircularProgressBase__root");
        const defaultCircularProgressSize = 24;

        expect(muiButton).toBeDisabled();
        expect(muiCircularProgress).toHaveStyle(`
        color: ${primaryColor};
        width: ${defaultCircularProgressSize}px;
        `);
        expect(screen.getByText("Test")).toBeInTheDocument();
    });

    it("should exist children with isLoading true and primary color", () => {
        props = {
            children: "Test",
            color: "primary",
            isLoading: true,
        };

        render(
            <TestThemeProvider>
                <ButtonBase {...props} />
            </TestThemeProvider>
        );

        const muiButton = screen.getByTestId("ButtonBase__root");
        const muiCircularProgress = screen.getByTestId("CircularProgressBase__root");
        const defaultCircularProgressSize = 24;

        expect(muiButton).toBeDisabled();
        expect(muiCircularProgress).toHaveStyle(`
        color: ${secondaryColor};
        width: ${defaultCircularProgressSize}px;
        `);
        expect(screen.getByText("Test")).toBeInTheDocument();
    });

    it("should exist children with isLoading true and circular progress small size", () => {
        props = {
            children: "Test",
            isLoading: true,
            size: "small",
        };

        render(
            <TestThemeProvider>
                <ButtonBase {...props} />
            </TestThemeProvider>
        );

        const muiButton = screen.getByTestId("ButtonBase__root");
        const muiCircularProgress = screen.getByTestId("CircularProgressBase__root");
        const smallCircularProgressSize = 20;

        expect(muiButton).toBeDisabled();
        expect(muiCircularProgress).toHaveStyle(`
        color: ${primaryColor};
        width: ${smallCircularProgressSize}px;
        `);
        expect(screen.getByText("Test")).toBeInTheDocument();
    });

    it("should exist children with isLoading true and circular progress medium size", () => {
        props = {
            children: "Test",
            isLoading: true,
            size: "medium",
        };

        render(
            <TestThemeProvider>
                <ButtonBase {...props} />
            </TestThemeProvider>
        );

        const muiButton = screen.getByTestId("ButtonBase__root");
        const muiCircularProgress = screen.getByTestId("CircularProgressBase__root");
        const mediumCircularProgressSize = 24;

        expect(muiButton).toBeDisabled();
        expect(muiCircularProgress).toHaveStyle(`
        color: ${primaryColor};
        width: ${mediumCircularProgressSize}px;
        `);
        expect(screen.getByText("Test")).toBeInTheDocument();
    });

    it("should exist children with isLoading true and circular progress large size", () => {
        props = {
            children: "Test",
            isLoading: true,
            size: "large",
        };

        render(
            <TestThemeProvider>
                <ButtonBase {...props} />
            </TestThemeProvider>
        );

        const muiButton = screen.getByTestId("ButtonBase__root");
        const muiCircularProgress = screen.getByTestId("CircularProgressBase__root");
        const largeCircularProgressSize = 28;

        expect(muiButton).toBeDisabled();
        expect(muiCircularProgress).toHaveStyle(`
        color: ${primaryColor};
        width: ${largeCircularProgressSize}px;
        `);
        expect(screen.getByText("Test")).toBeInTheDocument();
    });
});
