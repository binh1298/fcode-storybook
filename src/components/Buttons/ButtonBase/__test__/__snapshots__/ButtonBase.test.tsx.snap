// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonBase /> should match snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    style="border-radius: 50em;"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <div>
        Test
      </div>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
